### PR TITLE
Implement Goto Type Definition

### PR DESCRIPTION
**Pull request by mpickering**
The test fails for some reason locally but works in the editor so if someone knows how to fix it then please tell me.

### CODE REVIEW

File: src/Development/IDE/Spans/AtPoint.hs [hunk 1]
```diff
@@ -123,32 +153,47 @@ locationsAtPoint
...
+  -> Position
+  -> [SpanInfo]
+  -> m [Location]
+querySpanInfoAt getSpan IdeOptions{..} pos =
+    fmap (map srcSpanToLocation) . mapMaybeM getSpan . spansAtPoint pos
+
+-- | Given a 'Name' attempt to find the location where it is defined.
+nameToLocation :: Monad f => (Module -> f (Maybe (HieFile, String))) -> Name -> f (Maybe SrcSpan)
+nameToLocation getHieFile name =
+              case nameSrcSpan name of
```
Comment (cocreature) [inline src/Development/IDE/Spans/AtPoint.hs:176]: The indentation here seems a bit excessive.

File: test/exe/Main.hs [hunk 2]
```diff
@@ -1323,7 +1323,11 @@ findDefinitionAndHoverTests = let
 
   mkFindTests tests = testGroup "get"
     [ testGroup "definition" $ mapMaybe fst tests
-    , testGroup "hover"      $ mapMaybe snd tests ]
+    , testGroup "hover"      $ mapMaybe snd tests
+    , testGroup "type-definition" $ typeDefinitionTests ]
```
Comment (cocreature) [inline test/exe/Main.hs:1327]:
```suggestion
    , testGroup "type-definition" typeDefinitionTests ]
```

**Comment by cocreature** [comment]
This has a merge conflict in `AtPoint.hs`.

**Comment by mpickering** [comment]
I have no idea how the tests work at the moment as they don't even typecheck? I fixed that and then see this log:

```
HIE
  get
    type-definition
      Saturated data con:   ghcide version: 0.1.0 (GHC: 8.8.3) (PATH: /home/matt/ghcide-clean/dist-newstyle/build/x86_64-linux/ghc-8.8.3/ghcide-0.1.0/x/ghcide/build/ghcide/ghcide) (GIT hash: e231f7a0bc5f52bc3de1c5d4659274b7ba57d6d5)
Starting LSP server...
If you are seeing this in a terminal, you probably should have run ghcide WITHOUT the --lsp option!
 Started LSP server in 0.00s
[DEBUG] Set files of interest to: [NormalizedFilePath "/run/user/1000/extra-dir-58388217296644/GotoHover.hs"]
[DEBUG] Starting shakeRun (aborting the previous one took 0.00s)
[INFO] Opened text document: file:///run/user/1000/extra-dir-58388217296644/GotoHover.hs
[INFO] Files created or deleted: [(NormalizedFilePath "/run/user/1000/extra-dir-58388217296644/GotoHover.hs",True)]
[DEBUG] Starting shakeRun (aborting the previous one took 0.00s)
[DEBUG] Finishing shakeRun (took 0.00s, exception: AsyncCancelled)
[INFO] TypeDefinition request at position 19:21 in file: /run/user/1000/extra-dir-58388217296644/GotoHover.hs
[DEBUG] Finishing shakeRun (took 0.00s, exception: AsyncCancelled)
[DEBUG] Starting shakeRun (aborting the previous one took 0.00s)
[DEBUG] Running cradle /run/user/1000/extra-dir-58388217296644
"success"
"success"
"success"
[DEBUG] Finishing shakeRun (took 0.14s, completed)
```

But the server never sends a response to the `TypeDefinition` request.

At this point, I am not going to look into it further as it seems like a deep rabbit hole not really concerned with this patch.

**Comment by mpickering** [comment]
I have fixed the tests and identified two bugs in lsp-test in the process..

**Comment by mpickering** [comment]
CI is green now, this is ready to merge.

**Pull request merged**